iree/compiler/Dialect/HAL/Transforms/CSEVariableLoads.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+class CSEVariableLoadsPass
+    : public PassWrapper<CSEVariableLoadsPass, OperationPass<FuncOp>> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto funcOp = getOperation();
+
+    Block *entryBlock = &funcOp.body().front();
+    auto entryBlockBuilder = OpBuilder::atBlockBegin(entryBlock);
+
+    DenseMap<StringRef, SmallVector<VariableLoadOp, 4>>
+        loadOpsGroupedByVariable;
+    funcOp.walk([&](VariableLoadOp loadOp) {
```

Comment:
change this to just one walk on Operation * and `isa<VariableLoadOp>`/etc to add to the maps - then you're only walking the function once (instead of twice like here, or three times once you add the check for StoreIndirect, or four times when adding the check for calls/etc etc)